### PR TITLE
fix: fix some bugs by testing with ceresdb

### PR DESCRIPTION
**Pull request by chunshao90**
# Which issue does this PR close?

Closes #

# Rationale for this change
Fixed some bugs when testing with ceresdb clusters.
 

# What changes are included in this PR?
* Update shardTopology version when table is created or dropped
*  Refactor some error message
*  Refactor some code

# Are there any user-facing changes?
No.

# How does this change test
Manual testing.

### CODE REVIEW

File: server/cluster/cluster.go [hunk 1]
```diff
@@ -254,6 +254,98 @@ func (c *Cluster) updateTableCacheLocked(shardID uint32, schema *Schema, tablePb
...
+	}, nil
+}
+
+func (c *Cluster) createTableUpdateTopologyLocked(ctx context.Context, shardID uint32, schema *Schema, tablePb *clusterpb.Table) (*CreateTableResult, error) {
+	// Update shardTopology in storage.
+	shardTopologies, err := c.storage.ListShardTopologies(ctx, c.clusterID, []uint32{shardID})
+	if err != nil {
+		return nil, errors.WithMessage(err, "get shard topology from storage")
+	}
+	if len(shardTopologies) != 1 {
```
Comment (ZuLiangWang) [inline server/cluster/cluster.go:282]: It's weird to check length of shardTopologies, it's better to define `GetShardTopologyByID` and replace `ListShardTopologies`.

File: server/cluster/cluster.go [hunk 2]
```diff
@@ -254,6 +254,98 @@ func (c *Cluster) updateTableCacheLocked(shardID uint32, schema *Schema, tablePb
...
+	}, nil
+}
+
+func (c *Cluster) dropTableUpdateCacheLocked(ctx context.Context, shardID uint32, schema *Schema, table *Table) (*DropTableResult, error) {
+	// Update shardTopology in storage.
+	shardTopologies, err := c.storage.ListShardTopologies(ctx, c.clusterID, []uint32{shardID})
+	if err != nil {
+		return nil, errors.WithMessage(err, "get shard topology from storage")
+	}
+	if len(shardTopologies) != 1 {
```
Comment (ZuLiangWang) [inline server/cluster/cluster.go:314]: Same as above.

File: server/cluster/cluster.go [hunk 3]
```diff
@@ -254,6 +254,98 @@ func (c *Cluster) updateTableCacheLocked(shardID uint32, schema *Schema, tablePb
...
+		return nil, ErrGetShardTopology.WithCausef("shard has more than one shard topology, shardID:%d, shardTopologies:%v",
+			shardID, shardTopologies)
+	}
+
+	shardTopology := shardTopologies[0]
+	shardTopology.TableIds = append(shardTopology.TableIds, tablePb.GetId())
+	storageVersion := shardTopology.Version
```
Comment (ShiKaiWi) [inline server/cluster/cluster.go:289]: How about naming it `prevVersion`?

File: server/cluster/cluster.go [hunk 4]
```diff
@@ -254,6 +254,98 @@ func (c *Cluster) updateTableCacheLocked(shardID uint32, schema *Schema, tablePb
...
+	}
+	if len(shardTopologies) != 1 {
+		return nil, ErrGetShardTopology.WithCausef("shard has more than one shard topology, shardID:%d, shardTopologies:%v",
+			shardID, shardTopologies)
+	}
+
+	shardTopology := shardTopologies[0]
+	// Remove table in shardTopology.
+	for i, id := range shardTopology.TableIds {
+		if id == table.GetID() {
```
Comment (ShiKaiWi) [inline server/cluster/cluster.go:322]: Shall we check whether the table to remove exists in the topology?

File: server/cluster/types.go [hunk 5]
```diff
@@ -31,7 +31,15 @@ type NodeShard struct {
...
-	Table       *Table
+	Table        *Table
+	ShardVersion *ShardVersion
+}
+
+type DropTableResult struct {
+	ShardVersion *ShardVersion
+}
+
+type ShardVersion struct {
```
Comment (ZuLiangWang) [inline server/cluster/types.go:42]:
```suggestion
type ShardUpdateResult struct {
```

File: server/cluster/cluster.go [hunk 6]
```diff
@@ -254,6 +254,98 @@ func (c *Cluster) updateTableCacheLocked(shardID uint32, schema *Schema, tablePb
...
+	if err = c.storage.PutShardTopology(ctx, c.clusterID, storageVersion, shardTopology); err != nil {
+		return nil, errors.WithMessage(err, "put shard topology")
+	}
+
+	// Update tableCache in memory.
+	table := c.updateTableCacheLocked(shardID, schema, tablePb)
+	shardVersion, err := c.updateShardTopologyVersionLocked(shardID, storageVersion)
+	if err != nil {
+		return nil, errors.WithMessage(err, "update shard topology version")
+	}
```
Comment (ShiKaiWi) [inline server/cluster/cluster.go:300]: Shall we define a method called `createTableInCache` which does:
- Add the new table to the schema;
- Update the `shardCache`;

And a method called `dropTableInCache` may also be needed.

File: server/cluster/cluster.go [hunk 7]
```diff
@@ -254,6 +254,98 @@ func (c *Cluster) updateTableCacheLocked(shardID uint32, schema *Schema, tablePb
...
+	shardTopology.Version = oldVersion + 1
+	if err = c.storage.PutShardTopology(ctx, c.clusterID, oldVersion, shardTopology); err != nil {
+		return nil, errors.WithMessage(err, "put shard topology")
+	}
+
+	// Update tableCache in memory.
+	schema.dropTableLocked(table.GetName())
+	for _, s := range c.shardsCache {
+		s.dropTableLocked(table.GetID())
+	}
```
Comment (ShiKaiWi) [inline server/cluster/cluster.go:338]: Only the shard owning such table needs removing the table.

File: server/cluster/types.go (hunk 5 above)
Comment (ShiKaiWi) [inline server/cluster/types.go:42]: Maybe we should call this `ShardVersionUpdate`.

File: server/coordinator/procedure/storage_impl.go [hunk 8]
```diff
@@ -69,10 +69,10 @@ func (e EtcdStorageImpl) MarkDeleted(ctx context.Context, id uint64) error {
 
 func (e EtcdStorageImpl) List(ctx context.Context, batchSize int) ([]*Meta, error) {
 	metas := make([]*Meta, 0)
-	do := func(_ string, value []byte) error {
+	do := func(key string, value []byte) error {
 		meta := &Meta{}
 		if err := decode(meta, string(value)); err != nil {
-			return errors.WithMessage(err, "decode meta failed")
+			return errors.WithMessagef(err, "decode meta failed, key:%s, value:%s", key, string(value))
```
Comment (ShiKaiWi) [inline server/coordinator/procedure/storage_impl.go:75]: Just format value as %v, because it may not be a valid utf8-encoded bytes.

**Review by ShiKaiWi — APPROVED** [review]
LGTM

**Pull request merged**